File: java/FlowFieldsGen/src/main/java/in/mcxiv/animation/TimeLine.java
package in.mcxiv.animation;

import in.mcxiv.animation.events.StartEvent;
import in.mcxiv.animation.events.StopEvent;

import java.util.ArrayDeque;
import java.util.ArrayList;

public class TimeLine implements Actor {

    private final ArrayList<Event> events = new ArrayList<>();
    private final ArrayList<Actor> actors = new ArrayList<>();

    private double time = 0;

    public void addEvent(Event event) {
        events.add(event);
    }

    public void addEvent(Actor actor, double start, double stop) {
        addEvent(new StartEvent(actor, start));
        addEvent(new StopEvent(actor, stop));
    }

    @Override
    public void step(float delta) {
        time += delta;
        stepEvents(delta);
        for (Actor actor : actors) actor.step(delta);
    }

    private void stepEvents(float delta) {
        ArrayDeque<Event> consumedEvents = new ArrayDeque<>();
        for (Event event : events)
            if (event.hasTimeCome(time)) {
                stepEvent(event, delta);
                consumedEvents.add(event);
            }
        events.removeAll(consumedEvents); // TODO: welp, check if it's not removing two values of the same object? We might be reusing events in future.
    }

    private void stepEvent(Event event, float delta) {
        switch (event.getName()) {
            case "StartEvent":
                actors.add(event.getActor());
                event.start();
                break;
            case "StopEvent":
                actors.remove(event.getActor());
                event.stop();
                break;
        }
    }
}

File: static_web/FlowFieldsGen/js/animation/TimeLine.js
class TimeLine extends Actor {
  constructor() {
    super();
    this.events = [];
    this.actors = [];

    this.time = 0;
  }

  addSimpleEvent(event) {
    this.events.push(event);
  }

  addEvent(actor, start, stop) {
    this.addSimpleEvent(new StartEvent(actor, start));
    this.addSimpleEvent(new StopEvent(actor, stop));
  }

  step(delta) {
    this.time += delta;
    this.stepEvents(delta);
    this.actors.forEach(actor => actor.step(delta))
  }

  stepEvents(delta) {
    let consumedEvents = [];
    for (let i = 0; i < this.events.length; i++) {
      let event = this.events[i];
      if (event.hasTimeCome(this.time)) {
        this.stepEvent(event, delta);
        consumedEvents.push(this.events.indexOf(event));
      }
    }
    consumedEvents.forEach(value => this.events.splice(value, 1));
  }

  stepEvent(event, delta) {
    switch (event.getName()) {
      case "StartEvent":
        this.actors.push(event.getActor());
        event.start();
        break;
      case "StopEvent":
        this.actors.splice(this.actors.indexOf(event.getActor()), 1);
        event.stop();
        break;
    }
  }
}
